packages/@aws-cdk/aws-appmesh/lib/tls-certificate.ts

```diff
@@ -0,0 +1,192 @@
+import * as acm from '@aws-cdk/aws-certificatemanager';
+import { CfnVirtualGateway, CfnVirtualNode } from './appmesh.generated';
+
+/**
+ * Enum of supported TLS modes
+ */
+export enum TlsMode {
+  /**
+   * Only accept encrypted traffic
+   */
+  STRICT = 'STRICT',
+
+  /**
+   * Accept encrypted and plaintext traffic.
+   */
+  PERMISSIVE = 'PERMISSIVE',
+
+  /**
+   * TLS is disabled, only accept plaintext traffic.
+   */
+  DISABLED = 'DISABLED',
+}
+
+/**
+ * Redundant API shapes are represented here.
+ */
+export interface TlsCertificateConfig {
+  /**
+   * The CFN shape for a virtual gateway listener TLS certificate
+   */
+  readonly virtualGatewayListenerTlsCertificate: CfnVirtualGateway.VirtualGatewayListenerTlsCertificateProperty,
+
+  /**
+   * The CFN shape for a virtual node listener TLS certificate
+   */
+  readonly virtualNodeListenerTlsCertificate: CfnVirtualNode.ListenerTlsCertificateProperty,
```

Comment:
If you are 100% certain that `CfnVirtualGateway.VirtualGatewayListenerTlsCertificateProperty` and `CfnVirtualNode.ListenerTlsCertificateProperty` will never have different properties, you can use a single one here, thanks to TypeScript's structural typing.

See the discussion in the other PR: https://github.com/aws/aws-cdk/pull/11563#discussion_r532958694

Reply:
We're confident these structures will stay the same moving forward. I did it this way to adhere to the [design guidelines](https://github.com/aws/aws-cdk/blob/master/DESIGN_GUIDELINES.md#unions). Am I misinterpreting this design guide?

Reply:
I'm not suggesting using unions. I'm saying that, thanks to TypeScript's structural typing, because `CfnVirtualGateway.VirtualGatewayListenerTlsCertificateProperty` and `CfnVirtualNode.ListenerTlsCertificateProperty` are identical, you only need a single property here.

Like I said, take a look at the code that was pushed in #11563, specifically: https://github.com/aws/aws-cdk/blob/bfee58c702c31fb8e89cf99c8b6fb944ef6a96a4/packages/%40aws-cdk/aws-appmesh/lib/client-policy.ts#L13-L18 , and the comment thread https://github.com/aws/aws-cdk/pull/11563#discussion_r532958694 .